### PR TITLE
Add back the nethost archive.

### LINKED ISSUE #49026: Symbol and NetHosts Archives no longer produced
Both the symbols zip and nethost zips stopped getting published after we moved to the SFX SDK. If these are still needed for something, we should bring them back and add them to the nightly table.

cc: @NikolaMilosavljevic @jkoritzinsky

### PR DESCRIPTION
**Pull request by jkoritzinsky**
This archive makes it a bit easier for developers to pull down nethost in a non-.NET environment (without NuGet). We had this in 3.1, but it was lost at some point between then and now with all of the large infrastructure changes.

With https://github.com/dotnet/runtime/pull/59842, fixes #49026

cc: @vitek-karas @AaronRobinsonMSFT @elinor-fung should we backport this to 6.0 (aka do we feel that shipping the nethost archive in 6.0 is important enough to backport)

### CODE REVIEW

**Comment by AaronRobinsonMSFT** [comment]
I think this is .NET 6 worthy. It would be good to have a solid argument though. I think source-build is enough but make sure that is known.

File: src/installer/pkg/archives/dotnet-nethost.proj [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    </ItemGroup>
+
+    <Copy SourceFiles="@(FilesToPublish)"
+          DestinationFiles="%(FilesToPublish.Destination)" />
+  </Target>
+
+  <Target Name="PublishSymbolsToDisk">
+    <Error Condition="'$(SymbolsOutputPath)' == ''" Text="Publishing to disk requires the SymbolsOutputPath to be set to the root of the path to write to." />
+
+    <Copy SourceFiles="$(DotNetHostBinDir)\dotnet.pdb"
```
Comment (AaronRobinsonMSFT) [inline src/installer/pkg/archives/dotnet-nethost.proj:32]: We will also want the `nethost.pdb` so people can avoid suppressing linking warnings.

Comment (jkoritzinsky) [reply src/installer/pkg/archives/dotnet-nethost.proj]: I've updated the project to create the symbols archive and publish it in the same manner as the nethost archive.

File: src/installer/pkg/archives/dotnet-nethost.proj [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+
+  <Target Name="PublishToDisk">
+    <Error Condition="'$(OutputPath)' == ''" Text="Publishing to disk requires the OutputPath to be set to the root of the path to write to." />
+
+    <ItemGroup>
+      <FilesToPublish Include="$(DotNetHostBinDir)\$(LibPrefix)nethost$(LibSuffix)"
+            Destination="$(OutputPath)$(LibPrefix)nethost$(LibSuffix)" />
+      <FilesToPublish Include="$(DotNetHostBinDir)\$(LibPrefix)libnethost$(StaticLibSuffix)"
+            Destination="$(OutputPath)$(LibPrefix)libnethost$(StaticLibSuffix)" />
```
Comment (hoyosjs) [inline src/installer/pkg/archives/dotnet-nethost.proj:18]:
```suggestion
            Destination="$(OutputPath)$(LibPrefix)nethost$(StaticLibSuffix)" />
```

Previously in 3.1 this was just 

```
      <NetHostFiles Include="$(_NetHostOutputDir)nethost.h" />
      <NetHostFiles Include="$(_NetHostOutputDir)$(LibPrefix)nethost$(LibSuffix)" />
      <NetHostFiles Condition="'$(OSGroup)' == 'Windows_NT'" Include="$(_NetHostOutputDir)nethost.lib" />
```

Is there both a lib and a static lib?

Comment (elinor-fung) [reply src/installer/pkg/archives/dotnet-nethost.proj:18]: Yes, there is both. The static lib was added in 5.0. It should be `libnethost.a` on Unix and `libnethost.lib` on Windows, so I think this should be:
```suggestion
            Destination="$(OutputPath)$(StaticLibPrefix)nethost$(StaticLibSuffix)" />
```

File: src/installer/pkg/archives/dotnet-nethost.proj [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  </ItemGroup>
+
+
+  <Target Name="PublishToDisk">
+    <Error Condition="'$(OutputPath)' == ''" Text="Publishing to disk requires the OutputPath to be set to the root of the path to write to." />
+
+    <ItemGroup>
+      <FilesToPublish Include="$(DotNetHostBinDir)\$(LibPrefix)nethost$(LibSuffix)"
+            Destination="$(OutputPath)$(LibPrefix)nethost$(LibSuffix)" />
+      <FilesToPublish Include="$(DotNetHostBinDir)\$(LibPrefix)libnethost$(StaticLibSuffix)"
```
Comment (elinor-fung) [inline src/installer/pkg/archives/dotnet-nethost.proj:17]:
```suggestion
      <FilesToPublish Include="$(DotNetHostBinDir)\$(StaticLibPrefix)nethost$(StaticLibSuffix)"
```

File: src/installer/pkg/archives/dotnet-nethost.proj [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    </ItemGroup>
+
+    <Copy SourceFiles="@(FilesToPublish)"
+          DestinationFiles="%(FilesToPublish.Destination)" />
+  </Target>
+
+  <Target Name="PublishSymbolsToDisk">
+    <Error Condition="'$(SymbolsOutputPath)' == ''" Text="Publishing to disk requires the SymbolsOutputPath to be set to the root of the path to write to." />
+
+    <Copy SourceFiles="$(DotNetHostBinDir)\$(LibPrefix)nethost$(SymbolsSuffix);$(DotNetHostBinDir)\$(LibPrefix)libnethost$(SymbolsSuffix)"
```
Comment (jkoritzinsky) [inline src/installer/pkg/archives/dotnet-nethost.proj:33]:
```suggestion
    <Copy SourceFiles="$(DotNetHostBinDir)\$(LibPrefix)nethost$(SymbolsSuffix);$(DotNetHostBinDir)\$(StaticLibPrefix)nethost$(SymbolsSuffix)"
```

**Comment by vitek-karas** [comment]
Would be nice to have this for 6.0. But since this is fully backward compatible, people can use 7.0 builds just as well.

File: src/installer/pkg/archives/dotnet-nethost.proj [hunk 5]
```diff
@@ -0,0 +1,40 @@
+<Project Sdk="Microsoft.Build.NoTargets">
+  <PropertyGroup>
+    <CreateSymbolsArchive>true</CreateSymbolsArchive>
```
Comment (jkoritzinsky) [inline src/installer/pkg/archives/dotnet-nethost.proj:3]: Source-build doesn't strip the host binaries, so don't create a symbols archive when source-building.
```suggestion
    <CreateSymbolsArchive Condition="'$(DotNetBuildFromSource)' != 'true'">true</CreateSymbolsArchive>
```

**Comment by jkoritzinsky** [comment]
All CI failures are due to https://github.com/dotnet/runtime/issues/59908. This PR is ready for another round of review

File: src/installer/pkg/archives/dotnet-nethost.proj [hunk 6]
```diff
@@ -0,0 +1,40 @@
+<Project Sdk="Microsoft.Build.NoTargets">
+  <PropertyGroup>
+    <CreateSymbolsArchive Condition="'$(DotNetBuildFromSource)' != 'true'">true</CreateSymbolsArchive>
```
Comment (ViktorHofer) [inline src/installer/pkg/archives/dotnet-nethost.proj:3]: Isn't this needed for sourcebuild and why wouldn't it work for sourcebuild?

File: src/installer/pkg/archives/Directory.Build.props [hunk 7]
```diff
@@ -0,0 +1,7 @@
+<Project>
+  <PropertyGroup>
+    <EnsureRuntimeIdentifierSet>true</EnsureRuntimeIdentifierSet>
```
Comment (ViktorHofer) [inline src/installer/pkg/archives/Directory.Build.props:3]: Why is this needed?

File: src/installer/pkg/archives/dotnet-nethost.proj [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.DotNet.Build.Tasks.Archives" Version="$(MicrosoftDotNetBuildTasksArchivesVersion)" />
```
Comment (ViktorHofer) [inline src/installer/pkg/archives/dotnet-nethost.proj:7]: Curious, what's the entry point in that package? Who is responsible for invoking it? I don't see any of the targets below call into it (depend on a target in it).

File: src/installer/pkg/archives/dotnet-nethost.proj [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+  </ItemGroup>
+
+
```
Comment (ViktorHofer) [inline src/installer/pkg/archives/dotnet-nethost.proj:10]: nit: empty line

```suggestion
```

File: src/installer/pkg/archives/dotnet-nethost.proj (hunk 6 above)
Comment (jkoritzinsky) [reply src/installer/pkg/archives/dotnet-nethost.proj:3]: The source-build build for the hosts doesn't strip symbols, so there's no symbol file to archive.

File: src/installer/pkg/archives/Directory.Build.props (hunk 7 above)
Comment (jkoritzinsky) [reply src/installer/pkg/archives/Directory.Build.props:3]: The Archives package uses the RID to determine the default archive name.

File: src/installer/pkg/archives/dotnet-nethost.proj (hunk 6 above)
Comment (ViktorHofer) [reply src/installer/pkg/archives/dotnet-nethost.proj:3]: Ah we have a similar case in the System.IO.Ports native package: https://github.com/dotnet/runtime/blob/b9eb5acbde03e0b7ddda6a16443ef4490a19aced/src/libraries/System.IO.Ports/pkg/runtime.native.System.IO.Ports.props#L11-L12. I think it would make more sense to use that condition and add a comment here as well.

Comment (jkoritzinsky) [reply src/installer/pkg/archives/dotnet-nethost.proj:3]:
```suggestion
	<!-- When KeepNativeSymbols is set, debug symbols are kept in the .so files. So when it is true, we don't need a symbols archive. -->
    <CreateSymbolsArchive Condition="'$(KeepNativeSymbols)' != 'true'">true</CreateSymbolsArchive>
```

**Review by ViktorHofer — APPROVED** [review]
